### ISSUE TITLE
full tests fail in release mode

### ISSUE THREAD

**Issue by gnzlbg** [rust-dev-tools/rust-semverver#99]
`RUST_TEST_THREADS=1 cargo test --release` fails

**Comment by ibabushkin** [comment]
On 2019-03-26, gnzlbg wrote:
> `RUST_TEST_THREADS=1 cargo test --release` fails
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-dev-tools/rust-semverver/issues/99

I am aware of and working on this issue :)
@whentze and I are in the process of revisiting the harness for full
test cases, but a race condition still persists.

**Comment by whentze** [comment]
Some additional info:

- On this computer, the tests also fail without `--release`
- The tests that fail are `full::libc0` and `full::libc1`, but only if both are executed in the same test run. Either test always passes when run individually.
- The failures appear to be nondeterministic: sometimes, both will fail, but other times, only one of them will.
- When a test fails, the file it's supposed to output into is empty.

**Comment by gnzlbg** [comment]
I always run the tests with `RUST_TEST_THREADS=1` so that they don't run in parallel because I believe that when testing the same crate (both `libc0` and `libc1` work on the `libc` crate), both tests are executed simultaneously on the same directory. If you run them one after the other, that does not happen.

**Comment by gnzlbg** [comment]
I don't know the best way to attack that. It might be as simple as creating different directories for each full test, and running the whole test there. 

That would solve another issue with the current approach. Currently, the git diff relies on the test modifying the crate's directory. Ideally that shouldn't happen, e.g., once we run the tests in a different directory in the OUTPUT_DIR of the crate, we'll just do a diff between the expected output and the one produced, but the expect output wouldn't be modified by the test. This has been annoying in the past because sometimes one accidentally commits modifications to the expected output.

**Comment by whentze** [comment]
To be frank, I find the test setup in `full.rs` kind of clumsy anyway. Why is the new output even written to a file at all? We could just store the process' output in a buffer and compare it with the on-disk expected output.

The setup also means that, in order for the tests to pass, you need to have this git repository checked out. If you try to just pull a source tarball from github and run the tests in there, it will not work, which is unfortunate.

**Comment by ibabushkin** [comment]
On 2019-03-26, whentze wrote:
> To be frank, I find the test setup in `full.rs` kind of clumsy anyway. Why is the new output even written to a file at all? We could just store the process' output in a buffer and compare it with the on-disk expected output.
> 
> The setup also means that, in order for the tests to pass, you need to have this git repository checked out. If you try to just pull a source tarball from github and run the tests in there, it will not work, which is unfortunate.

I think it is very useful to keep the test output from the last run
around as I regularly find myself reading it multiple times and
searching in it etc. when debugging regressions for example.

I agree with the arguments agains using git's diffing functionality for
that. We should probably move to outputting to a temporary file (which
should be easy to find if necessary) and diffing it with the version
checked into git (and distributed with the source via other channels as
well).

**Comment by whentze** [comment]
How about this, then:

We store the output into a buffer and compare it (not with a diff algorithm, just direct strcmp) against the expected output. If they agree, the test passes and we do nothing.

If they do not compare equal, write the differing output to a file (somewhere in the target directory, perhaps? that is already gitignored and suitable for generated files). Devs can then diff that against the expected output using whatever tool they prefer.

**Comment by ibabushkin** [comment]
On 2019-03-26, whentze wrote:
> How about this, then:
> 
> We store the output into a buffer and compare it (not with a diff algorithm, just direct strcmp) against the expected output. If they agree, the test passes and we do nothing.
> 
> If they do not compare equal, write the differing output to a file (somewhere in the target directory, perhaps? that is already gitignored and suitable for generated files). Devs can then diff that against the expected output using whatever tool they prefer.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-dev-tools/rust-semverver/issues/99#issuecomment-476684670

That sounds good. I suggest the same file name as the expected output,
but in the directory below `target/` that mirrors the source directory
`tests/full_cases`. Does that make sense?

**Comment by gnzlbg** [comment]
Sounds good. 

I'd appreciate if on error either the `diff x y` command that failed, or at least the path to the test output in the target directory was printed to stderr to avoid having to manually scrap the output directory for the offending output file.

**Comment by whentze** [comment]
Sounds reasonable.

I'll try and write up a PR today.

**Comment by gnzlbg** [comment]
Awesome. @ibabushkin we should try and nag @rust-dev-tools (cc @Manishearth) until we get CI back up.

**Comment by whentze** [comment]
I've wondered for a while now where in the target directory the new files should go, but couldn't really find anything useful about the structure of that directory. I've come up with another, more pragmatic solution:

Put the new output in the `full_cases` directory as well, with the same filename but suffixed with `.new`.
We can just add those to gitignore, it's not a big deal.
This has the added benefit that the diff command can be very short, like so:
`git diff tests/full_cases/libc-0.2.28-0.2.31.linux*`

**Comment by gnzlbg** [comment]
You should query the `$OUT_DIR` environment variable, note that depending on which `--target` and build mode you use, you might end up with `target/i686-unknown-linux-gnu/release` instead of just `target/debug/...` 

So IMO these should go in `$OUT_DIR/full_cases/{case_name}` (e.g. where case_name is libc0, libc1, etc.).

**Comment by whentze** [comment]
So after playing around for a while I found this out: `cargo test` does not set `$OUT_DIR` unless the crate has a `build.rs`, so it doesn't work in our case (unless we want to add a dummy `build.rs`, which sounds like a bad idea). It doesn't set `CARGO_TARGET_DIR` either.

I'll just do the version I suggested in my last reply as I can't see any real downsides to it.

**Comment by gnzlbg** [comment]
> unless we want to add a dummy build.rs, which sounds like a bad idea

So I'd prefer that over modifying / creating files in the tests/full_cases directory, mainly because this will make `cargo clean` "just work" and because it will allow us to mount the source directory as read-only (which enables other things, like using `cross` for making sure that rust-semverver works on other targets). 

Having to add a `build.rs` just to get these environment variables set for tests feels wrong, but beyond that I don't see any big downsides, and i think I'll just rather fill in a cargo bug for it. I don't know how controversial changing cargo to always do this would be, but for our needs it would be enough if cargo would only set these variables when running `cargo test`, which doesn't sound that bad.

**Comment by whentze** [comment]
I'd tend to agree that having an immutable source directory is more important than not having a `build.rs`, so I've written up a PR for that version.

**Comment by ibabushkin** [comment]
Since we merged #104, this issue should be dealt with. I verified that on my local machine too, if someone notices anything fishy, please tell me.
